File: composeApp/src/androidMain/kotlin/com/roberto/goodbooks/ui/library/ShelfDetailViewModel.kt
package com.roberto.goodbooks.ui.library

import android.app.Application
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.roberto.goodbooks.GoodBooksApp
import com.roberto.goodbooks.db.Book
import com.roberto.goodbooks.db.Category
import com.roberto.goodbooks.domain.toBookItem
import com.roberto.goodbooks.network.models.BookItem
import kotlinx.coroutines.launch

class ShelfDetailViewModel(application: Application) : AndroidViewModel(application) {

    private val repository = getApplication<GoodBooksApp>().bookRepository

    // La estantería que estamos viendo
    var shelf by mutableStateOf<Category?>(null)
        private set

    // Los libros que hay DENTRO
    var booksInShelf by mutableStateOf<List<Book>>(emptyList())
        private set

    // Los libros DISPONIBLES para añadir (Todos - Los que ya están)
    var availableBooks by mutableStateOf<List<Book>>(emptyList())
        private set

    fun loadShelf(shelfId: Long) {
        viewModelScope.launch {
            // 1. Buscamos la info de la estantería
            val allShelves = repository.getAllShelves()
            shelf = allShelves.find { it.id == shelfId }

            // 2. Cargamos los libros que tiene dentro
            refreshBooks(shelfId)
        }
    }

    private suspend fun refreshBooks(shelfId: Long) {
        booksInShelf = repository.getBooksInShelf(shelfId)
    }

    // Prepara la lista de libros que se pueden añadir
    fun loadAvailableBooks() {
        viewModelScope.launch {
            val all = repository.getAllBooks()
            // Filtramos: Solo mostramos los que NO están ya en la estantería
            val existingIds = booksInShelf.map { it.gid }.toSet()
            availableBooks = all.filter { it.gid !in existingIds }
        }
    }

    fun addBookToShelf(book: Book) {
        val s = shelf ?: return
        viewModelScope.launch {
            repository.addBookToShelf(book.gid, s.id)
            refreshBooks(s.id)
            loadAvailableBooks() // Actualizamos disponibles
        }
    }

    fun removeBookFromShelf(book: Book) {
        val s = shelf ?: return
        viewModelScope.launch {
            repository.removeBookFromShelf(book.gid, s.id)
            refreshBooks(s.id)
        }
    }
}

File: iosApp/iosApp/ShelfDetailViewModel.swift
import Foundation
import Shared

@MainActor
class ShelfDetailViewModel: ObservableObject {
    
    private let repository: BookRepository
    let shelfId: Int64
    let shelfName: String // Guardamos el nombre para el título
    
    @Published var booksInShelf: [Book] = []
    @Published var availableBooks: [Book] = [] // Libros candidatos para añadir
    @Published var isLoading = false
    
    init(repository: BookRepository, shelfId: Int64, shelfName: String) {
        self.repository = repository
        self.shelfId = shelfId
        self.shelfName = shelfName
        loadBooks()
    }
    
    // Carga los libros que ya están dentro
    func loadBooks() {
        isLoading = true
        booksInShelf = repository.getBooksInShelf(shelfId: shelfId)
        isLoading = false
    }
    
    // Carga los libros que PODRÍAS añadir (Todos menos los que ya están)
    func loadAvailableBooks() {
        let allBooks = repository.getAllBooks()
        // Filtramos: Solo queremos los que NO estén ya en la lista booksInShelf
        // Usamos un Set de IDs para que la búsqueda sea rápida
        let existingIds = Set(booksInShelf.map { $0.gid })
        
        availableBooks = allBooks.filter { !existingIds.contains($0.gid) }
    }
    
    func addBook(_ book: Book) {
        repository.addBookToShelf(bookId: book.gid, shelfId: shelfId)
        loadBooks()          // Recargar la lista principal
        loadAvailableBooks() // Recargar la lista de candidatos (para quitar el que acabas de añadir)
    }
    
    func removeBook(at offsets: IndexSet) {
        for index in offsets {
            let book = booksInShelf[index]
            repository.removeBookFromShelf(bookGid: book.gid, shelfId: shelfId)
        }
        loadBooks()
    }
}
